src/main/java/org/truffleruby/core/string/FrozenStringLiterals.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core.string;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.truffleruby.RubyContext;
+import org.truffleruby.collections.ConcurrentOperations;
+import org.truffleruby.core.rope.Rope;
+
+import com.oracle.truffle.api.object.DynamicObject;
+
+public class FrozenStringLiterals {
+
+    private final RubyContext context;
+    private final Map<Rope, DynamicObject> strings = new ConcurrentHashMap<>();
```

Comment:
Does this need to be concurrent? The only way to add a frozen string literal is through parsing and that's all done behind a lock, isn't it?

Reply:
The `require` lock is not a global lock but a per-file lock, so parsing can happen in parallel.

Reply:
Thanks for clarifying.